src/abstract-interpretation/data-frame/domain.ts

```diff
@@ -0,0 +1,324 @@
+ * */
+export interface DataFrameDomain {
+    colnames: ColNamesDomain,
+    cols:     IntervalDomain,
+    rows:     IntervalDomain
+}
+
+/**
+ * The bottom element (least element) of the data frame shape domain representing no possible value, mapping the columns names to {@link ColNamesBottom}
+ * and the number of columns and rows to {@link IntervalBottom}.
+ */
+export const DataFrameBottom: DataFrameDomain = {
+	colnames: ColNamesBottom,
+	cols:     IntervalBottom,
+	rows:     IntervalBottom
+};
+
+/**
+ * The top element (greatest element) of the data frame shape domain representing all possible value, mapping the columns names to {@link ColNamesTop}
+ * and the number of columns and rows to {@link IntervalTop}.
+ */
+export const DataFrameTop: DataFrameDomain = {
+	colnames: ColNamesTop,
+	cols:     IntervalTop,
+	rows:     IntervalTop
+};
+
+/**
+ * The data frame shape state domain representing possible memory states, mapping AST node IDs to data frame shape values of the {@link DataFrameDomain}.
+ */
+export type DataFrameStateDomain = Map<NodeId, DataFrameDomain>;
+
+/** Checks if two abstract values of the column names domain are equal. */
+export function equalColNames(set1: ColNamesDomain, set2: ColNamesDomain): boolean {
+	return set1 === set2 || (set1 !== ColNamesTop && set2 !== ColNamesTop && setEquals(new Set(set1), new Set(set2)));
+}
+
+/** Checks if two abstract values of the column names domain are ordered according to the partial ordering of the column names lattice. */
+export function leqColNames(set1: ColNamesDomain, set2: ColNamesDomain): boolean {
+	return set2 === ColNamesTop || (set1 !== ColNamesTop && new Set(set1).isSubsetOf(new Set(set2)));
+}
+
+/** Joins two abstract values of the columns names domain according to the column names lattice by creating the least upper bound (LUB). */
+export function joinColNames(set1: ColNamesDomain, set2: ColNamesDomain, maxColNames: number = MaxColNames): ColNamesDomain {
+	if(set1 === ColNamesTop || set2 === ColNamesTop) {
+		return ColNamesTop;
+	}
+	const join = Array.from(new Set(set1).union(new Set(set2)));
+
+	return join.length > maxColNames ? ColNamesTop : join;
+}
+
+/** Meets two abstract values of the columns names domain according to the column names lattice by creating the greatest lower bound (GLB). */
+export function meetColNames(set1: ColNamesDomain, set2: ColNamesDomain): ColNamesDomain {
+	if(set1 === ColNamesTop) {
+		return set2;
+	} else if(set2 === ColNamesTop) {
+		return set1;
+	} else {
+		return Array.from(new Set(set1).intersection(new Set(set2)));
+	}
+}
+
+/** Subtracts an abstract value from another abstract value of the column names domain by performing a set minus. */
+export function subtractColNames(set1: ColNamesDomain, set2: ColNamesDomain): ColNamesDomain {
+	if(set1 === ColNamesTop) {
+		return ColNamesTop;
+	} else if(set2 === ColNamesTop) {
+		return set1;
+	} else {
+		return Array.from(new Set(set1).difference(new Set(set2)));
```

Comment:
if column names would imply order, this would destroy that. If they do not, why use arrays in the first place?

Reply:
Ok this is fine, but i guess then #1760 is scheduled to happen _after_ this PR?

Reply:
Although afterward, I would be interested whether we get performance benefits just from changing the structure and avoiding the recasting on operations